relay-server/src/services/buffer/envelope_provider/sqlite.rs

```diff
@@ -0,0 +1,626 @@
+            max_disk_size: config.spool_envelopes_max_disk_size(),
+        }
+    }
+
+    /// Initializes the envelope provider.
+    ///
+    /// Retrieves the project key pairs from the envelope store and creates
+    /// an initialization state.
+    pub async fn initialize(&self) -> InitializationState {
+        match self.envelope_store.project_key_pairs().await {
+            Ok(project_key_pairs) => InitializationState::new(project_key_pairs),
+            Err(error) => {
+                relay_log::error!(
+                    error = &error as &dyn Error,
+                    "failed to initialize the sqlite stack provider"
+                );
+                InitializationState::empty()
+            }
+        }
+    }
+
+    /// Pushes an envelope to the provider for the given project key pair.
+    ///
+    /// If the spool threshold is exceeded, it may trigger spooling to disk.
+    pub async fn push(
+        &mut self,
+        project_key_pair: ProjectKeyPair,
+        envelope: Box<Envelope>,
+    ) -> Result<(), SqliteEnvelopeProviderError> {
+        if self.above_spool_threshold() {
+            self.spool_to_disk().await?;
+        }
+
+        self.envelopes
+            .entry(project_key_pair)
+            .or_default()
+            .push(envelope);
+
+        self.buffered_envelopes_size += 1;
+
+        Ok(())
+    }
+
+    /// Peeks at the next envelope for the given project key pair without removing it.
+    ///
+    /// If no envelope is in the buffer, it will be loaded from disk and a reference will be
+    /// returned.
+    pub async fn peek(
+        &mut self,
+        project_key_pair: ProjectKeyPair,
+    ) -> Result<Option<&Envelope>, SqliteEnvelopeProviderError> {
+        if self.is_empty(project_key_pair) {
+            let envelopes = self.unspool_from_disk(project_key_pair, 1).await?;
+            if envelopes.is_empty() {
+                return Ok(None);
+            }
+
+            self.buffered_envelopes_size += envelopes.len() as u64;
+            self.envelopes
+                .entry(project_key_pair)
+                .or_default()
+                .extend(envelopes);
+        }
+
+        Ok(self
+            .envelopes
+            .get(&project_key_pair)
+            .and_then(|e| e.last().map(Box::as_ref)))
+    }
+
+    /// Pops and returns the next envelope for the given project key pair.
+    ///
+    /// If no envelope is in the buffer, it will be loaded from disk.
+    pub async fn pop(
+        &mut self,
+        project_key_pair: ProjectKeyPair,
+    ) -> Result<Option<Box<Envelope>>, SqliteEnvelopeProviderError> {
+        let envelope = self
+            .envelopes
+            .get_mut(&project_key_pair)
+            .and_then(|envelopes| envelopes.pop());
+
+        // If we have no more envelopes to pop from the buffer, we want to remove the project key
+        // pair to free up some memory.
+        if self.is_empty(project_key_pair) {
```

Comment:
Does `is_empty` mean "memory empty" or "memory + disk empty"? Because if it's just memory-empty, we should not remove the stack from the hash map.